File: Data Structures & Algorithms/surrounded-regions/submission-1.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        nrows, ncols = len(board), len(board[0])
        def dfs(row, col):
            if 0 <= row < nrows and 0 <= col < ncols:
                if board[row][col] == "O":
                    board[row][col] = "#"
                    for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                        nr, nc = row + dr, col + dc
                        dfs(nr, nc)
            
        for row in range(nrows):
            if board[row][0] == "O":
                dfs(row, 0)
            if board[row][ncols - 1] == "O":
                dfs(row, ncols - 1)
        
        for col in range(1, ncols - 1):
            if board[0][col] == "O":
                dfs(0, col)
            if board[nrows - 1][col] == "O":
                dfs(nrows - 1, col)
        

        for row in range(nrows):
            for col in range(ncols):
                if board[row][col] == "#":
                    board[row][col] = "O"
                else:
                    board[row][col] = "X"
        
        

File: Data Structures & Algorithms/island-perimeter/submission-1.cs
public class Solution {
    public int IslandPerimeter(int[][] grid) {
        var (nrows, ncols) = (grid.Length, grid[0].Length);
        bool[,] visited = new bool[nrows, ncols];
        for (int i = 0; i < nrows; i++) {
            for (int j = 0; j < ncols; j++) {
                if (grid[i][j] == 1) {
                    return dfs(i, j, grid, visited);
                }
            }
        }
        return 0;
    }

    public int dfs(int row, int col, int[][] grid, bool[,] visited) {
        if (visited[row, col]) {
            // skip if visited
            return 0;
        }

        visited[row, col] = true;
        var perimeter = 0;
        var (nrows, ncols) = (grid.Length, grid[0].Length);
        (int dr, int dc)[] directions =
        {
            (-1, 0),
            (1, 0),
            (0, -1),
            (0, 1)
        };

        foreach (var (dr, dc) in directions) {
            var (nr, nc) = (row + dr, col + dc);

            if (nr < 0 || nr >= nrows || nc < 0 || nc >= ncols) {
                // we hit a wall so incrmeent perimeter
                perimeter++;
            }
            else if (grid[nr][nc] == 0) {
                // we hit water
                perimeter++;
            }
            else {
                perimeter += dfs(nr, nc, grid, visited);
            }
        }

        return perimeter;
    }
}